web/auth_test.go

```diff
@@ -0,0 +1,204 @@
+		expectedPamResult:   true,
+		expectedGroupResult: true,
+		expectedResult:      true,
+		description:         "good user with correct password should pass PAM, pass group check, and should validate",
+	},
+	testCase{
+		user:                testUsers["gooduser"],
+		testPassword:        "",
+		testGroup:           adminGroup,
+		expectedPamResult:   false,
+		expectedGroupResult: true,
+		expectedResult:      false,
+		description:         "good user with empty password should fail PAM, pass group check, and should not validate.",
+	},
+	testCase{
+		user:                testUsers["nonrootuser"],
+		testPassword:        testUsers["nonrootuser"].password,
+		testGroup:           adminGroup,
+		expectedPamResult:   true,
+		expectedGroupResult: false,
+		expectedResult:      false,
+		description:         "nonroot user should pass PAM, fail authentication, and should not validate.",
+	},
+	testCase{
+		user:                testUsers["oldbutgooduser"],
+		testPassword:        testUsers["oldbutgooduser"].password,
+		testGroup:           adminGroup,
+		expectedPamResult:   false,
+		expectedGroupResult: true,
+		expectedResult:      false,
+		description:         "user with expired login should pass PAM, fail group check, and should not validate.",
+	},
+}
+
+
+func TestMain(m *testing.M) {
+	if 0 != os.Geteuid() {
+		glog.Infof("Must be root to run integration tests. Exiting (no tests run).")
+		os.Exit(0)
+	}
+	err := CreateTestUsers()
+	if err != nil {
+		glog.Errorf("Error creating test user: %s\n", err)
+	}
+	glog.Infof("Running tests - some errors expected from negative tests.\n")
+	result := m.Run()
+	glog.Infof("Removing test users\n")
+	RemoveTestUsers()
+	glog.Infof("Exiting\n")
+	os.Exit(result)
+}
+
+func CreateTestUsers() error {
+	for name, u := range (testUsers) {
+		err := createTestUser(name, &u)
+		if err != nil {
+			glog.Errorf("Error creating user %s: %s\n", name, err)
+			return err
+		}
+		glog.V(2).Infof("Created user %s. user = %v\n", name, u)
+	}
+	return nil
+}
+
+
+func createTestUser(name string, userobj *testUser) error {
+	testUserName := userobj.username
+	if _, err := user.Lookup(testUserName); err == nil {
+		return ErrTestUserExists
+	}
+	encryptedPassword := crypt(userobj.password, testUserSalt)
+	cmdName := "sudo"
+	args := []string{"useradd", userobj.username, "-p", encryptedPassword, "-G", userobj.group }
+	if userobj.expired {
+		args = append(args, "-e", "1970-01-01")
+	}
+	command := exec.Command(cmdName, args...)
+	glog.V(2).Infof("Creating test user %s: command is %v\n", testUserName, command)
+	output, cmderr := command.CombinedOutput()
+	if nil != cmderr {
+		glog.Errorf("Error creating testuser: %s\n%s\n", cmderr, output)
+	} else {
+		createdUsers[name] = *userobj
+	}
+	return cmderr
+}
+
+func RemoveTestUsers() {
+	for _, user := range (createdUsers) {
+		err := RemoveTestUser(user.username)
+		if err != nil {
+			glog.Infof("Error deleting user %s: %s\n", user.username, err)
```

Comment:
Fixed. This was an outgrowth of the same philosophy I used on the non-root user - I didn't think errors in testing should mark the code as broken. As above, I can see the value of viewing the code and test environment as a whole.